File: tests/fixtures/multi_format_corpus/code/realistic/go/inventory.go
// Package inventory implements a small in-memory inventory ledger.
//
// It mirrors the Python / TypeScript / Rust fixtures: items keyed by
// name with integer quantities, restock + guarded withdrawal, and
// aggregate reporting. Idiomatic Go for the code-embedder lane.
package inventory

import (
	"errors"
	"fmt"
)

// ErrOutOfStock is returned when a withdrawal exceeds the quantity on hand.
var ErrOutOfStock = errors.New("out of stock")

// Item is a single inventory line: a name and the quantity in stock.
type Item struct {
	Name     string
	Quantity int
}

// Inventory is a collection of items keyed by name.
type Inventory struct {
	items map[string]*Item
}

// New returns an empty, ready-to-use Inventory.
func New() *Inventory {
	return &Inventory{items: make(map[string]*Item)}
}

// Restock adds amount units of name, creating the item if needed.
func (inv *Inventory) Restock(name string, amount int) error {
	if amount < 0 {
		return errors.New("amount must be non-negative")
	}
	if item, ok := inv.items[name]; ok {
		item.Quantity += amount
		return nil
	}
	inv.items[name] = &Item{Name: name, Quantity: amount}
	return nil
}

// Withdraw removes amount units of name and returns the new quantity.
//
// It returns a wrapped ErrOutOfStock if too few units are available.
func (inv *Inventory) Withdraw(name string, amount int) (int, error) {
	item, ok := inv.items[name]
	if !ok {
		return 0, fmt.Errorf("unknown item %q", name)
	}
	if item.Quantity < amount {
		return 0, fmt.Errorf("%s: %w", name, ErrOutOfStock)
	}
	item.Quantity -= amount
	return item.Quantity, nil
}

// TotalUnits returns the sum of all quantities across every item.
func (inv *Inventory) TotalUnits() int {
	total := 0
	for _, item := range inv.items {
		total += item.Quantity
	}
	return total
}

File: tests/fixtures/multi_format_corpus/code/realistic/rust/inventory.rs
//! A small in-memory inventory ledger.
//!
//! Mirrors the Python / TypeScript / Go fixtures: items keyed by name
//! with integer quantities, restock + guarded withdrawal, and aggregate
//! reporting. Idiomatic Rust for the code-embedder lane.

use std::collections::HashMap;

/// Why a withdrawal failed.
#[derive(Debug, PartialEq, Eq)]
pub enum WithdrawError {
    /// No item is registered under the requested name.
    Unknown,
    /// The item exists but does not have enough units on hand.
    OutOfStock { have: u32, need: u32 },
}

/// A single inventory line: a name and the quantity in stock.
#[derive(Debug, Clone)]
pub struct Item {
    pub name: String,
    pub quantity: u32,
}

/// A collection of items keyed by name.
#[derive(Debug, Default)]
pub struct Inventory {
    items: HashMap<String, Item>,
}

impl Inventory {
    /// Create an empty inventory.
    pub fn new() -> Self {
        Inventory {
            items: HashMap::new(),
        }
    }

    /// Add `amount` units of `name`, creating the item if needed.
    pub fn restock(&mut self, name: &str, amount: u32) {
        match self.items.get_mut(name) {
            Some(item) => item.quantity += amount,
            None => {
                self.items.insert(
                    name.to_string(),
                    Item {
                        name: name.to_string(),
                        quantity: amount,
                    },
                );
            }
        }
    }

    /// Remove `amount` units of `name`, returning the new quantity.
    pub fn withdraw(&mut self, name: &str, amount: u32) -> Result<u32, WithdrawError> {
        let item = self.items.get_mut(name).ok_or(WithdrawError::Unknown)?;
        if item.quantity < amount {
            return Err(WithdrawError::OutOfStock {
                have: item.quantity,
                need: amount,
            });
        }
        item.quantity -= amount;
        Ok(item.quantity)
    }

    /// Sum of all quantities across every item.
    pub fn total_units(&self) -> u32 {
        self.items.values().map(|item| item.quantity).sum()
    }
}
